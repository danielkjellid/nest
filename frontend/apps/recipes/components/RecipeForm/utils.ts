import {
  type RecipeCreateIn,
  type RecipeEditIn,
  type IngredientItem,
  type RecipeIngredientItemRecord,
  type RecipeStepRecord,
  type Step,
  type IngredientGroupItem,
  type RecipeIngredientItemGroupRecord,
} from '../../../../types'

import {
  type Recipe,
  type IngredientItem as FormIngredientItem,
  type Step as FormStep,
  type IngredientItemGroup as FromIngredientItemGroup,
} from './types'

const makeIngredientItemType = (
  ingredientItem: RecipeIngredientItemRecord | FormIngredientItem
): IngredientItem => ({
  id: 'id' in ingredientItem ? ingredientItem.id : undefined,
  ingredient: ingredientItem.ingredient.id.toString(),
  portionQuantity: ingredientItem.portionQuantity.toString(),
  portionQuantityUnit: ingredientItem.portionQuantityUnit.id.toString(),
  additionalInfo: ingredientItem.additionalInfo || undefined,
})

const makeStepType = (step: FormStep | RecipeStepRecord, index: number): Step => ({
  id: 'id' in step ? step.id : undefined,
  instruction: step.instruction,
  stepType: step.stepType,
  duration: step.duration,
  number: index + 1,
  ingredientItems: step.ingredientItems.map((ingredientItem: any) =>
    makeIngredientItemType(ingredientItem)
  ),
})

const makeIngredientItemGroupType = (
  ingredientGroup: FromIngredientItemGroup | RecipeIngredientItemGroupRecord
): IngredientGroupItem => ({
  id: 'id' in ingredientGroup ? ingredientGroup.id : undefined,
  title: ingredientGroup.title,
  ordering: ingredientGroup.ordering,
  ingredientItems: ingredientGroup.ingredientItems.map((ingredientItem) =>
    makeIngredientItemType(ingredientItem)
  ),
})

const makePayload = (recipeData: Recipe): RecipeCreateIn | RecipeEditIn => ({
  baseRecipe: { ...recipeData.baseRecipe },
  steps: [...recipeData.steps.map((step, index) => makeStepType(step, index))],
  ingredientItemGroups: [
    ...recipeData.ingredientItemGroups.map((ingredientGroup) =>
      makeIngredientItemGroupType(ingredientGroup)
    ),
  ],
})

export { makePayload }
