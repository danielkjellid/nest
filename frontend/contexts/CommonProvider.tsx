import React, { createContext, useContext } from 'react'

export interface Home {
  id: number
  address: string
  numResidents: number
  numWeeksRecipeRotation: number
  weeklyBudget: number
  isActive: boolean
}

export interface User {
  id: number
  email: string
  firstName: string
  lastName: string
  fullName: string
  isActive: boolean
  isStaff: boolean
  isSuperuser: boolean
  home?: Home
}

export interface Config {
  isProduction: boolean
}

export interface CommonContextType {
  config: Config
  currentUser: User
  currentHome: Home | null
  availableHomes: Home[]
  setCurrentHome(home: Home): void
}

const CommonContext = createContext<CommonContextType | null>(null)

export function useCommonContext() {
  const ctx = useContext(CommonContext)

  if (!ctx) {
    throw new Error(
      'useCommonContext hook was called outside of context, make sure your app is wrapped with CommonProvider component'
    )
  }

  return ctx
}

interface CommonProviderProps extends CommonContextType {
  children: React.ReactNode
}

export function CommonProvider({
  config,
  currentUser,
  currentHome,
  availableHomes,
  setCurrentHome,
  children,
}: CommonProviderProps) {
  return (
    <CommonContext.Provider
      value={{ config, currentUser, currentHome, availableHomes, setCurrentHome }}
    >
      {children}
    </CommonContext.Provider>
  )
}
