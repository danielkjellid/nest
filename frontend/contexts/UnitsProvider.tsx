import { createContext, useContext, useMemo, type ReactNode } from 'react'

import { type UnitListOut } from '../types'

export interface UnitOption {
  label: string
  value: string
}

export interface UnitsContextType {
  units: UnitListOut[] | undefined
  unitsOptions: UnitOption[] | undefined
}

const UnitsContext = createContext<UnitsContextType | undefined>(undefined)

export function useUnits() {
  const ctx = useContext(UnitsContext)

  if (!ctx) {
    throw new Error(
      'useUnits hook was called outside of context, make sure your app is wrapped with UnitsProvider component'
    )
  }

  return ctx
}

interface UnitsProviderProps extends UnitsContextType {
  children: ReactNode
}

export function UnitsProvider({ units, unitsOptions, children }: UnitsProviderProps) {
  const value = useMemo(() => ({ units, unitsOptions }), [units, unitsOptions])
  return <UnitsContext.Provider value={value}>{children}</UnitsContext.Provider>
}
