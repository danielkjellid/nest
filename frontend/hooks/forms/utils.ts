import { type FormElementObj, type FormElement } from '../../components/Form/types'

import { type SchemaFormElementObj, type SchemaFormElement } from './types'

type K = keyof FormElementObj
type KS = keyof SchemaFormElementObj
type Mapping = Record<KS, K>

const MAPPING: Mapping = {
  title: 'title',
  type: 'type',
  enum: 'enum',
  'x-helpText': 'helpText',
  'x-component': 'component',
  'x-defaultValue': 'defaultValue',
  'x-placeholder': 'placeholder',
  'x-hiddenLabel': 'hiddenLabel',
  'x-colSpan': 'colSpan',
  'x-section': 'section',
  'x-order': 'order',
  'x-min': 'min',
  'x-max': 'max',
}

const convertProperty = (property: SchemaFormElementObj): FormElementObj => {
  const mapping = (key: KS): K => MAPPING[key] as K
  const convertedProperty: Partial<FormElementObj> = {}

  Object.entries(property).map(([k, v]) => {
    const mappedKey = mapping(k as KS)
    convertedProperty[mappedKey] = v
  })

  return convertedProperty as FormElementObj
}

const convertSchemaElemToFormElem = (schemaProperties: SchemaFormElement): FormElement => {
  const convertedProperties: Partial<FormElement> = {}

  Object.entries(schemaProperties).map(([key, property]) => {
    convertedProperties[key] = convertProperty(property)
  })

  return convertedProperties as FormElement
}

export { convertSchemaElemToFormElem }
