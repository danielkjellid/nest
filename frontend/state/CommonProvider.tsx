import React, { createContext, useContext } from 'react'

interface Home {
  id: number
  address: string
  numResidents: number
  numWeeksRecipeRotation: number
  weeklyBudget: number
  isActive: boolean
}

interface User {
  id: number
  email: string
  firstName: string
  lastName: string
  fullName: string
  isActive: boolean
  isStaff: boolean
  isSuperuser: boolean
  home?: Home
}

interface Config {
  isProduction: boolean
}

export interface CommonContextType {
  config: Config
  currentUser: User
  currentHome: Home
  availableHomes: Home[]
  setCurrentHome(home: Home): void
}

const CommonContext = createContext<CommonContextType | null>(null)

export function useCommonContext() {
  const ctx = useContext(CommonContext)

  if (!ctx) {
    throw new Error(
      'useCommonContext hook was called outside of context, make sure your app is wrapped with CommonProvider component'
    )
  }

  return ctx
}

interface CommonProviderProps extends CommonContextType {
  children: React.ReactNode
}

export function CommonProvider({
  config,
  currentUser,
  currentHome,
  availableHomes,
  setCurrentHome,
  children,
}: CommonProviderProps) {
  return (
    <CommonContext.Provider
      value={{ config, currentUser, currentHome, availableHomes, setCurrentHome }}
    >
      {children}
    </CommonContext.Provider>
  )
}
