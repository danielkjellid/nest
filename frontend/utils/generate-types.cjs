/* eslint-disable */
const OpenAPI = require('openapi-typescript-codegen')
const fs = require('fs')
const path = require('path')

const outputPath = './frontend/types/generated/dist'

const generateTypesFromSchema = () => {
  fs.readFile(path.join(__dirname, '../../schema.json'), (err, file) => {
    OpenAPI.generate({
      input: JSON.parse(file.toString()),
      output: outputPath,
      exportCore: false,
      useOptions: true,
    })
  })
}

if (require.main === module) {
  generateTypesFromSchema()
}

module.exports = {
  generateTypesFromSchema,
}
