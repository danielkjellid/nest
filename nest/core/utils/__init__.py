from .dates import format_date, format_datetime, format_time
from .humps import camelize, decamelize, is_camelcase, is_snakecase
from .requests import get_remote_request_ip, get_remote_request_user

__all__ = [
    "camelize",
    "decamelize",
    "is_camelcase",
    "is_snakecase",
    "format_date",
    "format_time",
    "format_datetime",
    "get_remote_request_ip",
    "get_remote_request_user",
]
