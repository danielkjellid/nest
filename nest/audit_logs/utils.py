from decimal import Decimal
from typing import Any, Iterable

from django.db.models import Model
from django.utils.encoding import smart_str


def calculate_models_diff(
    old: Model | None, new: Model | None, fields: Iterable[str]
) -> dict[str, tuple[str | None, str | None]] | None:
    """
    Calculates the differences between two model instances. One of the
    instances may be None (i.e., a newly created model or deleted model). This
    will cause all fields with a value to have changed (from None).
    """

    if not (old is None or isinstance(old, Model)):
        raise TypeError("The supplied old instance is not a valid model instance.")
    if not (new is None or isinstance(new, Model)):
        raise TypeError("The supplied new instance is not a valid model instance.")

    diff = {}

    for field in fields:
        old_value = smart_str(getattr(old, field, None), strings_only=True)
        new_value = smart_str(getattr(new, field, None), strings_only=True)

        if old_value != new_value:
            if isinstance(old_value, Decimal) and isinstance(new_value, str):
                if old_value == Decimal(new_value):
                    continue

            elif isinstance(new_value, Decimal) and isinstance(old_value, str):
                if Decimal(old_value) == new_value:
                    continue

            diff[field] = (smart_str(old_value), smart_str(new_value))

    if not diff:
        diff = None  # type: ignore

    return diff


def format_changes_display(
    *, changes: dict[str, tuple[Any | None, Any | None]]
) -> list[str]:
    formatted_changes = [
        f"{key} changed from {old_value} to {new_value}"
        for key, (old_value, new_value) in changes.items()
    ]

    return formatted_changes
