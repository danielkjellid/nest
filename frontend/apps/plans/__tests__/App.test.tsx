import { render, screen } from '@testing-library/react'
import { describe, expect, it } from 'vitest'

import { createCommonContextTestData } from '../../../contexts/__tests__/utils'
import { CommonProvider } from '../../../contexts/CommonProvider'
import PlansApp from '../App'

describe('Test PageApp app', () => {
  it('empty state provided when currentHome is null and availableHomes is empty', () => {
    const { config, currentUser } = createCommonContextTestData()
    render(
      <CommonProvider
        currentHome={null}
        availableHomes={[]}
        config={config}
        currentUser={currentUser}
        setCurrentHome={() => console.log('Set home')}
      >
        <PlansApp />
      </CommonProvider>
    )

    expect(screen.getByText(/You are currently not assigned any homes/i)).toBeInTheDocument()
  })

  it('empty state matches snapshot', () => {
    const { config, currentUser } = createCommonContextTestData()

    const { asFragment } = render(
      <CommonProvider
        currentHome={null}
        availableHomes={[]}
        config={config}
        currentUser={currentUser}
        setCurrentHome={() => console.log('Set home')}
      >
        <PlansApp />
      </CommonProvider>
    )

    expect(asFragment()).toMatchSnapshot()
  })

  it('normal state provided when currentHome is not null and availableHomes is empty', () => {
    const { config, currentHome, currentUser } = createCommonContextTestData()
    render(
      <CommonProvider
        currentHome={currentHome}
        availableHomes={[]}
        config={config}
        currentUser={currentUser}
        setCurrentHome={() => console.log('Set home')}
      >
        <PlansApp />
      </CommonProvider>
    )

    expect(screen.getByText(/Plans/i)).toBeInTheDocument()
  })
})
