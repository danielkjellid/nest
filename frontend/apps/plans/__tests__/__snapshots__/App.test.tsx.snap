// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test PageApp app > empty state matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="mantine-1j5d86u undefined border-2 border-dashed rounded-md flex items-center justify-center p-8"
  >
    <div
      class="mantine-15ha46x text-center space-y-2"
    >
      <svg
        class="w-12 h-12 mx-auto"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="1.3"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12 12m-9 0a9 9 0 1 0 18 0a9 9 0 1 0 -18 0"
        />
        <path
          d="M9 10l.01 0"
        />
        <path
          d="M15 10l.01 0"
        />
        <path
          d="M9.5 15.25a3.5 3.5 0 0 1 5 0"
        />
      </svg>
      <h1
        class="mantine-Text-root mantine-Title-root mantine-174z13q"
      >
        You are currently not assigned any homes
      </h1>
      <div
        class="mantine-Text-root mantine-1d564l0"
      >
        Please contact an admin.
      </div>
    </div>
  </div>
</DocumentFragment>
`;
